### DISCUSSION TITLE
Parse JSONB/JSON columns when querying

### DISCUSSION THREAD

**Discussion by theelderbeever** [piccolo-orm/piccolo#466] [Q&A] [answered]
Is there a way to automatically parse JSONB/JSON data columns when performing a `Table.select().run()`? It seems like a useful feature instead of having to write the code to do it with every query.

**Comment by dantownsend** [comment, marked as answer]
@theelderbeever Yes, this should do the job:

```python
await RecordingStudio.select().output(load_json=True)
```

**Reply by XCanG** [reply]
How to do the same thing when you use `.raw()` method? Because it doesn't have `.output` method.

**Reply by sinisaos** [reply]
@XCanG Without adding an `output` method to the `Raw` class this is currently not possible. It would be something like this (you can also create this class somewhere in your code without changing the source code).
```python
from __future__ import annotations

from collections.abc import Sequence
from typing import TYPE_CHECKING, Optional, TypeVar

from piccolo.engine.base import BaseBatch
from piccolo.query.base import Query
from piccolo.query.mixins import OutputDelegate
from piccolo.querystring import QueryString

if TYPE_CHECKING:  # pragma: no cover
    from piccolo.table import Table


class Raw(Query):
    __slots__ = ("querystring", "output_delegate")

    def __init__(
        self,
        table: type[Table],
        querystring: QueryString = QueryString(""),
        **kwargs,
    ):
        super().__init__(table, **kwargs)
        self.querystring = querystring
        self.output_delegate = OutputDelegate()

    def output(
        self: Self,
        *,
        as_list: bool = False,
        as_json: bool = False,
        load_json: bool = False,
        nested: bool = False,
    ) -> Self:
        self.output_delegate.output(
            as_list=as_list,
            as_json=as_json,
            load_json=load_json,
            nested=nested,
        )
        return self

    async def batch(
        self,
        batch_size: Optional[int] = None,
        node: Optional[str] = None,
        **kwargs,
    ) -> BaseBatch:
        if batch_size:
            kwargs.update(batch_size=batch_size)
        if node:
            kwargs.update(node=node)
        return await self.table._meta.db.batch(self, **kwargs)

    @property
    def default_querystrings(self) -> Sequence[QueryString]:
        return [self.querystring]


Self = TypeVar("Self", bound=Raw)
```


Without that change, you only have the option of manually parsing the column. Something like this would work
```python
data = await Studio.raw("SELECT * FROM studio")
for item in data:
     item["facilities"] = json.loads(item["facilities"]) # <-- parse your column here
```

**Reply by dantownsend** [reply]
Yeah, it's tricky with raw queries. We don't know know which ones are the JSON columns, unless we can somehow get that info from asyncpg.

**Reply by dantownsend** [reply]
If the user used Piccolo's `Cast` function to explicitly tell us something is a `JSON` column, then maybe we could do it.

**Reply by XCanG** [reply]
I didn't quite like the suggestion because manually describing fields doesn't help for type hints, so instead I relied on Pydantic models and made fix like this:
```py
from typing import get_origin
import orjson
from pydantic import BaseModel

def unwrap_jsonb[T: BaseModel](model: type[T], row: dict, /) -> T:
    """
    Resolve Piccolo issue with `.raw()` method returning JSONB as strings

    ```
    {"array_field": '[""]'}
    ```
    """

    filtered_origins: set = {list, tuple, Sequence, dict}
    fields: set[str] = {k for k, v in model.model_fields.items() if get_origin(v.annotation) in filtered_origins}

    return model.model_validate({k: orjson.loads(v) if k in fields else v for k, v in row.items()})
```

**Reply by XCanG** [reply]
Regarding this topic I encountered similar issue with Django, when it comes to JSONB. For Django to execute raw SQL you use `cursor` and as a return type you get [`cursor.description`](https://docs.djangoproject.com/en/5.2/topics/db/sql/#:~:text=for%20col%20in-,cursor.description,-%5D%0A%20%20%20%20return%20%5B)

When looking at implementation of this method it's code for PostgreSQL is in `django/db/backends/postgresql/introspection.py`:
```py
def get_table_description(self, cursor, table_name):
    """
    Return a description of the table with the DB-API cursor.description
    interface.
    """
    # Query the pg_catalog tables as cursor.description does not reliably
    # return the nullable property and information_schema.columns does not
    # contain details of materialized views.
    cursor.execute(
        """
        SELECT
            a.attname AS column_name,
            NOT (a.attnotnull OR (t.typtype = 'd' AND t.typnotnull)) AS is_nullable,
            pg_get_expr(ad.adbin, ad.adrelid) AS column_default,
            CASE WHEN collname = 'default' THEN NULL ELSE collname END AS collation,
            a.attidentity != '' AS is_autofield,
            col_description(a.attrelid, a.attnum) AS column_comment
        FROM pg_attribute a
        LEFT JOIN pg_attrdef ad ON a.attrelid = ad.adrelid AND a.attnum = ad.adnum
        LEFT JOIN pg_collation co ON a.attcollation = co.oid
        JOIN pg_type t ON a.atttypid = t.oid
        JOIN pg_class c ON a.attrelid = c.oid
        JOIN pg_namespace n ON c.relnamespace = n.oid
        WHERE c.relkind IN ('f', 'm', 'p', 'r', 'v')
            AND c.relname = %s
            AND n.nspname NOT IN ('pg_catalog', 'pg_toast')
            AND pg_catalog.pg_table_is_visible(c.oid)
    """,
        [table_name],
    )
    field_map = {line[0]: line[1:] for line in cursor.fetchall()}
    cursor.execute(
        "SELECT * FROM %s LIMIT 1" % self.connection.ops.quote_name(table_name)
    )
    return [
        FieldInfo(
            line.name,
            line.type_code,
            # display_size is always None on psycopg2.
            line.internal_size if line.display_size is None else line.display_size,
            line.internal_size,
            line.precision,
            line.scale,
            *field_map[line.name],
        )
        for line in cursor.description
    ]
```

Then in the code you could use something like this:
```py
def cursor_value_to_python(v: Any, tc: int, /) -> Any:
    """
    Map DB type to Python type

    * 3802 = JSONB
    """

    if tc == 3802:  # ? jsonb
        return json.loads(v) if v is not None else None

    return v  # ? Other types


def cursor_fetchall(cursor: CursorWrapper, /) -> Sequence[dict[str, Any]]:
    """
    Fetch all rows and return them in format of iterable with dict (where key = column name)
    """

    return [
        {col.name: cursor_value_to_python(val, col.type_code) for col, val in zip(cursor.description or [], row)}
        for row in cursor.fetchall()
    ]
```

**Reply by sinisaos** [reply]
Sorry to both of you, but I still don't see the point of complex introspection to find JSON columns (to automatically deserialize JSON in response) when we can add a simple method to the `Raw` class and get the same result. Something like [this branch](https://github.com/piccolo-orm/piccolo/compare/master...sinisaos:piccolo:deserialize_raw_json_response). If that's not an option, we can always subclass the `Raw` class and make `CustomRaw` in the application code. Something like this works without problems.

<details>
<summary>Simple app example</summary>

```python
import asyncio
import uvicorn
from typing import Any, Self

from fastapi import FastAPI
from fastapi.responses import JSONResponse
from piccolo.columns import JSONB, Varchar
from piccolo.table import Table
from piccolo.query.mixins import OutputDelegate
from piccolo.query.methods import Raw
from piccolo.querystring import QueryString

from piccolo.engine.sqlite import SQLiteEngine


DB = SQLiteEngine()


# custom Raw class
class CustomRaw(Raw):
    __slots__ = ("output_delegate",)

    def __init__(
        self,
        table: type[Table],
        querystring: QueryString = QueryString(""),
        **kwargs,
    ):
        super().__init__(table, querystring, **kwargs)
        self.output_delegate = OutputDelegate()

    def deserialize_json(self: Self, *, load_json: bool = False) -> Self:
        self.output_delegate.output(load_json=load_json)
        return self


# Tables example
class Studio(Table, db=DB):
    name = Varchar()
    facilities = JSONB()

    # custom raw classmethod
    @classmethod
    def custom_raw(cls, sql: str, *args: Any) -> CustomRaw:
        return CustomRaw(table=cls, querystring=QueryString(sql, *args))


app = FastAPI()


@app.get("/")
async def index() -> JSONResponse:
    data = await Studio.custom_raw("SELECT * FROM studio").deserialize_json(
        load_json=True
    )
    return JSONResponse({"data": data})


async def main():
    await Studio.create_table(if_not_exists=True)


if __name__ == "__main__":
    asyncio.run(main())

    uvicorn.run(app, host="127.0.0.1", port=8000)
```
</details>

Again, I apologize to both of you if I'm wrong about this.

**Reply by dantownsend** [reply]
@sinisaos When Piccolo gets the rows from the database, it uses the Piccolo table the query was run on to figure out which are the `JSON` columns. I think the reason your PR works is because of this line:

https://github.com/piccolo-orm/piccolo/blob/5139ec0644675bc3b195d46de2d50571cfa9e4c7/piccolo/query/base.py#L82

The raw query is being run on the `RecordingStudio` table, which has `JSON` columns defined on it, which is why Piccolo is able to figure out what the `JSON` columns are.

If instead you did this, it probably wouldn't work:

```python
await Band.raw(
    "SELECT facilities, facilities_b FROM recording_studio"
).deserialize_json(load_json=True)
```

The challenge with raw queries is they could be doing pretty much anything, so it's hard to know which values are JSON or not.

Using `Cast` *might* work, because at least that is explicitly telling Piccolo that something is `JSON`. But that would still be tricky to make work.

Probably the best way of making it work is to use asyncpg codecs:

https://magicstack.github.io/asyncpg/current/usage.html#example-automatic-json-conversion

**Reply by sinisaos** [reply]
@dantownsend Sorry, I hadn't thought of it that way, but I don't understand why would someone run a raw query on a different table than the table the raw query was written for? The asyncpg codecs (we can put it [here](https://github.com/piccolo-orm/piccolo/blob/5139ec0644675bc3b195d46de2d50571cfa9e4c7/piccolo/engine/postgres.py#L524)) will only work if we explicit type cast the json column in query (so we go back to the original problem of detecting the json column) like this
```python
await Band.raw("SELECT id, name, facilities::json FROM studio")
```
Without that, the json column is not deserialized.

**Reply by dantownsend** [reply]
>  I hadn't thought of it that way, but I don't understand why would someone run a raw query on a different table than the table the raw query was written for?

Yeah, I know it sounds weird, but the raw query could potentially do anything.

> The asyncpg codecs (we can put it [here](https://github.com/piccolo-orm/piccolo/blob/5139ec0644675bc3b195d46de2d50571cfa9e4c7/piccolo/engine/postgres.py#L524)) will only work if we explicit type cast the json column in query (so we go back to the original problem of detecting the json column)

Hmm, interesting. I still think that's an OK solution, but it would require quite a bit of refactoring.

We could somehow make it easy for the user to register their own asyncpg codecs, and then have that as a solution - that way we wouldn't have to do much refactoring.

**Reply by XCanG** [reply]
> @dantownsend Sorry, I hadn't thought of it that way, but I don't understand why would someone run a raw query on a different table than the table the raw query was written for? The asyncpg codecs (we can put it [here](https://github.com/piccolo-orm/piccolo/blob/5139ec0644675bc3b195d46de2d50571cfa9e4c7/piccolo/engine/postgres.py#L524)) will only work if we explicit type cast the json column in query (so we go back to the original problem of detecting the json column) like this
> ```python
> await Band.raw("SELECT id, name, facilities::json FROM studio")
> ```
> Without that, the json column is not deserialized.

I can answer you that, because sometimes you just needed more data from other tables and so you join them in a way you ended up utilizing JSON/JSONB. for example I returning model of product that joined 5 topmost reviews, and so I have pydantic model like:

```py
class ProductResponse(Product): # original model
    reviews: Sequence[Review]
```
and so it become JSONB field, e. g. 
```sql
SELECT p.*, (
    SELECT jsonb_agg(r)
    FROM reviews r
    WHERE r.product_id = p.id
    ORDER BY r.likes DESC
    LIMIT 5
) AS "reviews"
FROM product p
WHERE p.id = {};
```
of course it's a simple example, but everything starts from there.

**Reply by sinisaos** [reply]
@XCanG Thanks for the example, but I think it has more to do with how you deserialize the raw query result than Piccolo. Converter like this works (at least in my case)
```python
def jsonb_converter(data: dict[str, Any]) -> dict[str, Any]:
    for key, value in data.items():
        try:
            if isinstance(value, str) and json.loads(value):
                data[key] = json.loads(value)
        except ValueError:
            data[key] = value
    return data
```
That converter works with 
```python
@app.get("/")
async def index() -> JSONResponse:
    rows = await Band.raw("SELECT * FROM studio")
    return JSONResponse({"rows": [jsonb_converter(row) for row in rows]})
```
Also works with `jsonb_agg`
```python
@app.get("/")
async def index() -> list[ProductResponse]:
    rows = await Product.raw(
        """
        SELECT p.*, (
            SELECT jsonb_agg(r)
            FROM review r
            WHERE r.product_id = p.id
        ) AS "reviews"
        FROM product p
        """
    )
    return [ProductResponse(**jsonb_converter(row)) for row in rows]
```
Result is:

<img width="1285" height="475" alt="raw_response" src="https://github.com/user-attachments/assets/4065fa81-b795-40e5-837f-b502ce233793" />

**Reply by XCanG** [reply]
@sinisaos This solution is even more riskier as it add possibilities to inject object into arbitrary strings, consider you have text field called `description: str` and with your code user input something like:
```py
>>> import json
>>> {"description": json.loads('"something"')}
{'description': 'something'}
>>> {"description": json.loads('["something"]')}
{'description': ['something']}
>>> {"description": json.loads('["something", false]')}
{'description': ['something', False]}
>>> {"description": json.loads('{"something": false}')}
{'description': {'something': False}}
>>> {"description": json.loads('0')}
{'description': 0}
>>> {"description": json.loads('Infinity')}
{'description': inf}
>>> {"description": json.loads('1e303')}
{'description': 1e+303}
>>> {"description": json.loads('null')}
{'description': None}
```
...
In some cases it will break models and produce uncertain errors.

And when it come to numbers, specifically BigInt's, it is known issue to parse them as numbers in other places, so large numbers typically sent as strings, with your code it will blindly convert them into integers and then frontend silently accept it but will parse wrongly.

JS example (considering it loading JSON from backend:
```js
JSON.parse(`{"bigint": 123456789012345678901}`)
// result:
{bigint: 123456789012345680000}
```

**Reply by sinisaos** [reply]
@XCanG OK, then forget about my answers. Sorry.
